File: core/src/util/timerange.cpp
/***

  Olive - Non-Linear Video Editor
  Copyright (C) 2023 Olive Studios LLC
  Modifications Copyright (C) 2025 mikesolar

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.

***/

#include "util/timerange.h"

#include <algorithm>
#include <cmath>
#include <utility>

#include "util/timecodefunctions.h"

namespace olive::core::internal
{

TimeRange::TimeRange(const Rational &in, const Rational &out)
	: in_(in)
	, out_(out)
{
	normalize();
}

const Rational &TimeRange::in() const
{
	return in_;
}

const Rational &TimeRange::out() const
{
	return out_;
}

const Rational &TimeRange::length() const
{
	return length_;
}

void TimeRange::set_in(const Rational &in)
{
	in_ = in;
	normalize();
}

void TimeRange::set_out(const Rational &out)
{
	out_ = out;
	normalize();
}

void TimeRange::set_range(const Rational &in, const Rational &out)
{
	in_ = in;
	out_ = out;
	normalize();
}

bool TimeRange::operator==(const TimeRange &r) const
{
	return in() == r.in() && out() == r.out();
}

bool TimeRange::operator!=(const TimeRange &r) const
{
	return in() != r.in() || out() != r.out();
}

bool TimeRange::overlaps_with(const TimeRange &a, bool in_inclusive,
							 bool out_inclusive) const
{
	bool doesnt_overlap_in = (in_inclusive) ? (a.out() < in()) :
											  (a.out() <= in());

	bool doesnt_overlap_out = (out_inclusive) ? (a.in() > out()) :
												(a.in() >= out());

	return !doesnt_overlap_in && !doesnt_overlap_out;
}

TimeRange TimeRange::combined(const TimeRange &a) const
{
	return combine(a, *this);
}

bool TimeRange::contains(const TimeRange &compare, bool in_inclusive,
						 bool out_inclusive) const
{
	bool contains_in = (in_inclusive) ? (compare.in() >= in()) :
										(compare.in() > in());

	bool contains_out = (out_inclusive) ? (compare.out() <= out()) :
										  (compare.out() < out());

	return contains_in && contains_out;
}

bool TimeRange::contains(const Rational &r) const
{
	return r >= in_ && r < out_;
}

TimeRange TimeRange::combine(const TimeRange &a, const TimeRange &b)
{
	return TimeRange(std::min(a.in(), b.in()), std::max(a.out(), b.out()));
}

TimeRange TimeRange::intersected(const TimeRange &a) const
{
	return intersect(a, *this);
}

TimeRange TimeRange::intersect(const TimeRange &a, const TimeRange &b)
{
	return TimeRange(std::max(a.in(), b.in()), std::min(a.out(), b.out()));
}

TimeRange TimeRange::operator+(const Rational &rhs) const
{
	TimeRange answer(*this);
	answer += rhs;
	return answer;
}

TimeRange TimeRange::operator-(const Rational &rhs) const
{
	TimeRange answer(*this);
	answer -= rhs;
	return answer;
}

const TimeRange &TimeRange::operator+=(const Rational &rhs)
{
	set_range(in_ + rhs, out_ + rhs);

	return *this;
}

const TimeRange &TimeRange::operator-=(const Rational &rhs)
{
	set_range(in_ - rhs, out_ - rhs);

	return *this;
}

std::list<TimeRange> TimeRange::split(const int &chunk_size) const
{
	std::list<TimeRange> split_ranges;

	int start_time =
		std::floor(this->in().to_double() / static_cast<double>(chunk_size)) *
		chunk_size;
	int end_time =
		std::ceil(this->out().to_double() / static_cast<double>(chunk_size)) *
		chunk_size;

	for (int i = start_time; i < end_time; i += chunk_size) {
		split_ranges.push_back(
			TimeRange(std::max(this->in(), Rational(i)),
					  std::min(this->out(), Rational(i + chunk_size))));
	}

	return split_ranges;
}

void TimeRange::normalize()
{
	// If `out` is earlier than `in`, swap them
	if (out_ < in_) {
		std::swap(out_, in_);
	}

	// Calculate length
	if (out_ == RATIONAL_MIN || out_ == RATIONAL_MAX || in_ == RATIONAL_MIN ||
		in_ == RATIONAL_MAX) {
		length_ = Rational::na_n;
	} else {
		length_ = out_ - in_;
	}
}

void TimeRangeList::insert(const TimeRangeList &list_to_add)
{
	for (auto it = list_to_add.cbegin(); it != list_to_add.cend(); it++) {
		insert(*it);
	}
}

void TimeRangeList::insert(TimeRange range_to_add)
{
	// See if list contains this range
	if (contains(range_to_add)) {
		return;
	}

	// Does not contain range, so we'll almost certainly be adding it in some way
	for (auto it = array_.begin(); it != array_.end();) {
		const TimeRange &compare = *it;

		if (compare.overlaps_with(range_to_add)) {
			range_to_add = TimeRange::combine(range_to_add, compare);
			it = array_.erase(it);
		} else {
			it++;
		}
	}

	array_.push_back(range_to_add);
}

void TimeRangeList::remove(const TimeRange &remove)
{
	util_remove(&array_, remove);
}

void TimeRangeList::remove(const TimeRangeList &list)
{
	for (const TimeRange &r : list) {
		remove(r);
	}
}

bool TimeRangeList::contains(const TimeRange &range, bool in_inclusive,
							 bool out_inclusive) const
{
	for (int i = 0; i < size(); i++) {
		if (array_.at(i).contains(range, in_inclusive, out_inclusive)) {
			return true;
		}
	}

	return false;
}

void TimeRangeList::shift(const Rational &diff)
{
	for (int i = 0; i < array_.size(); i++) {
		array_[i] += diff;
	}
}

void TimeRangeList::trim_in(const Rational &diff)
{
	// Re-do list since we want to handle overlaps
	TimeRangeList temp = *this;

	clear();

	for (auto it = temp.array_.begin(); it != temp.array_.end(); it++) {
		TimeRange &r = *it;
		r.set_in(r.in() + diff);
		insert(r);
	}
}

void TimeRangeList::trim_out(const Rational &diff)
{
	// Re-do list since we want to handle overlaps
	TimeRangeList temp = *this;

	clear();

	for (auto it = temp.array_.begin(); it != temp.array_.end(); it++) {
		TimeRange &r = *it;
		r.set_out(r.out() + diff);
		insert(r);
	}
}

TimeRangeList TimeRangeList::intersects(const TimeRange &range) const
{
	TimeRangeList intersect_list;

	for (int i = 0; i < size(); i++) {
		const TimeRange &compare = array_.at(i);

		if (compare.out() <= range.in() || compare.in() >= range.out()) {
			// No intersect
			continue;
		} else {
			// Crop the time range to the range and add it to the list
			TimeRange cropped(std::max(range.in(), compare.in()),
							  std::min(range.out(), compare.out()));

			intersect_list.insert(cropped);
		}
	}

	return intersect_list;
}

TimeRangeListFrameIterator::TimeRangeListFrameIterator()
	: TimeRangeListFrameIterator(TimeRangeList(), Rational::na_n)
{
}

TimeRangeListFrameIterator::TimeRangeListFrameIterator(
	const TimeRangeList &list, const Rational &timebase)
	: list_(list)
	, timebase_(timebase)
	, range_index_(-1)
	, size_(-1)
	, frame_index_(0)
	, custom_range_(false)
{
	if (!list_.isEmpty() && timebase_.isNull()) {
		std::cerr
			<< "TimeRangeListFrameIterator created with null timebase but non-empty list, this will likely lead to infinite loops"
			<< std::endl;
	}

	update_index_if_necessary();
}

Rational TimeRangeListFrameIterator::snap(const Rational &r) const
{
	return Timecode::snap_time_to_timebase(r, timebase_, Timecode::k_floor);
}

bool TimeRangeListFrameIterator::get_next(Rational *out)
{
	if (!has_next()) {
		return false;
	}

	// Output current value
	*out = current_;

	// Determine next value by adding timebase
	current_ += timebase_;

	// If this time is outside the current range, jump to the next one
	update_index_if_necessary();

	// Increment frame index
	frame_index_++;

	return true;
}

bool TimeRangeListFrameIterator::has_next() const
{
	return range_index_ < list_.size();
}

int TimeRangeListFrameIterator::size()
{
	if (size_ == -1) {
		// Size isn't calculated automatically for optimization, so we'll calculate it now
		size_ = 0;

		for (const TimeRange &range : list_) {
			Rational start = snap(range.in());
			Rational end = Timecode::snap_time_to_timebase(
				range.out(), timebase_, Timecode::k_floor);

			if (end == range.out()) {
				end -= timebase_;
			}

			int64_t start_ts = Timecode::time_to_timestamp(start, timebase_);
			int64_t end_ts = Timecode::time_to_timestamp(end, timebase_);

			size_ += 1 + (end_ts - start_ts);
		}
	}

	return size_;
}

void TimeRangeListFrameIterator::update_index_if_necessary()
{
	while (range_index_ < list_.size() &&
		   (range_index_ == -1 || current_ >= list_.at(range_index_).out())) {
		range_index_++;

		if (range_index_ < list_.size()) {
			current_ = snap(list_.at(range_index_).in());
		}
	}
}

}

File: src/oakcore-rs/src/timerange.rs
// Oak Video Editor - Non-Linear Video Editor
// Copyright (C) 2026 Oak Team
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

//! Time ranges and normalized range lists (oakcore `TimeRange` /
//! `TimeRangeList` equivalents).

use crate::rational::{self, Rational};

/// Half-open time range [in, out) — mirrors `olive::core::TimeRange`.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Default)]
pub struct TimeRange {
    in_: Rational,
    out: Rational,
}

impl TimeRange {
    /// Construct and normalize (C++ ctor calls `normalize()`: if `out <
    /// in` the two are swapped). The doc comment on the skeleton claimed
    /// normalization is not performed; matching C++ takes precedence.
    pub fn new(in_: Rational, out: Rational) -> TimeRange {
        let mut r = TimeRange { in_, out };
        r.normalize();
        r
    }

    /// Inclusive start.
    pub fn in_(&self) -> Rational {
        self.in_
    }

    /// Exclusive end.
    pub fn out(&self) -> Rational {
        self.out
    }

    /// `out - in`. When either endpoint is a `RATIONAL_MIN/MAX`
    /// sentinel the subtraction propagates NaN, matching C++ which
    /// stores the same sentinel value for `length_`.
    pub fn length(&self) -> Rational {
        self.out - self.in_
    }

    /// True when `t` lies in [in, out).
    pub fn contains(&self, t: Rational) -> bool {
        t >= self.in_ && t < self.out
    }

    /// True when `self` contains `compare`, honoring inclusivity of the
    /// in/out edges (C++ `TimeRange::contains(TimeRange)`).
    fn contains_range(
        &self,
        compare: &TimeRange,
        in_inclusive: bool,
        out_inclusive: bool,
    ) -> bool {
        let contains_in = if in_inclusive {
            compare.in_ >= self.in_
        } else {
            compare.in_ > self.in_
        };
        let contains_out = if out_inclusive {
            compare.out <= self.out
        } else {
            compare.out < self.out
        };
        contains_in && contains_out
    }

    /// True when `self` and `a` overlap, honoring edge inclusivity
    /// (C++ `TimeRange::overlaps_with`).
    fn overlaps_with(&self, a: &TimeRange, in_inclusive: bool, out_inclusive: bool) -> bool {
        let does_not_overlap_in = if in_inclusive {
            a.out < self.in_
        } else {
            a.out <= self.in_
        };
        let does_not_overlap_out = if out_inclusive {
            a.in_ > self.out
        } else {
            a.in_ >= self.out
        };
        !does_not_overlap_in && !does_not_overlap_out
    }

    /// Intersection; empty when disjoint (C++ `intersected`).
    ///
    /// Note: C++ normalizes the result, so disjoint inputs produce a
    /// swapped (in > out) range rather than an "empty" marker; we match
    /// that bit-for-bit.
    pub fn intersected(&self, other: &TimeRange) -> TimeRange {
        TimeRange::new(
            std::cmp::max(self.in_, other.in_),
            std::cmp::min(self.out, other.out),
        )
    }

    /// Union that also merges touching ranges (C++ `combined`).
    pub fn combined(&self, other: &TimeRange) -> TimeRange {
        TimeRange::new(
            std::cmp::min(self.in_, other.in_),
            std::cmp::max(self.out, other.out),
        )
    }

    /// C++ `set_in` + `normalize`.
    fn set_in(&mut self, in_: Rational) {
        self.in_ = in_;
        self.normalize();
    }

    /// C++ `set_out` + `normalize`.
    fn set_out(&mut self, out: Rational) {
        self.out = out;
        self.normalize();
    }

    /// C++ `normalize`: swap if `out < in`.
    fn normalize(&mut self) {
        if self.out < self.in_ {
            std::mem::swap(&mut self.out, &mut self.in_);
        }
    }
}

/// Normalized (sorted, non-overlapping) list of ranges — mirrors
/// `olive::core::TimeRangeList` including its merge-on-insert and
/// subtraction semantics.
#[derive(Clone, Debug, Default)]
pub struct TimeRangeList {
    ranges: Vec<TimeRange>,
}

impl TimeRangeList {
    /// Empty list.
    pub fn new() -> Self {
        TimeRangeList { ranges: Vec::new() }
    }

    /// True when any element fully contains `range` (C++
    /// `TimeRangeList::contains`, inclusive edges).
    fn contains_range(&self, range: &TimeRange) -> bool {
        self.ranges
            .iter()
            .any(|r| r.contains_range(range, true, true))
    }

    /// Insert a range, merging overlaps and touching neighbors
    /// (C++ `insert(TimeRange)`).
    pub fn insert(&mut self, range: TimeRange) {
        // If the list already fully contains this range, nothing to do.
        if self.contains_range(&range) {
            return;
        }

        let mut range = range;
        let mut i = 0;
        while i < self.ranges.len() {
            let compare = self.ranges[i];
            if compare.overlaps_with(&range, true, true) {
                range = compare.combined(&range);
                self.ranges.remove(i);
            } else {
                i += 1;
            }
        }

        self.ranges.push(range);
    }

    /// Subtract a range (C++ `remove`, via `util_remove`).
    pub fn remove(&mut self, range: TimeRange) {
        let mut additions: Vec<TimeRange> = Vec::new();

        let mut i = 0;
        while i < self.ranges.len() {
            let compare = self.ranges[i];

            if range.contains_range(&compare, true, true) {
                // The removal range entirely encompasses this element.
                self.ranges.remove(i);
            } else if compare.contains_range(&range, false, false) {
                // The removal range is strictly inside this element:
                // split it into two.
                let mut new_range = compare;
                new_range.set_in(range.out);
                let mut trimmed = compare;
                trimmed.set_out(range.in_);
                self.ranges[i] = trimmed;
                additions.push(new_range);
                break;
            } else {
                if compare.in_ < range.in_ && compare.out > range.in_ {
                    // This element's out overlaps the range's in: trim it.
                    self.ranges[i].set_out(range.in_);
                } else if compare.in_ < range.out && compare.out > range.out {
                    // This element's in overlaps the range's out: trim it.
                    self.ranges[i].set_in(range.out);
                }
                i += 1;
            }
        }

        self.ranges.extend(additions);
    }

    /// Sorted ranges view.
    pub fn ranges(&self) -> &[TimeRange] {
        &self.ranges
    }

    /// True when the list has no ranges.
    pub fn is_empty(&self) -> bool {
        self.ranges.is_empty()
    }

    /// Total covered duration (sum of each range's length).
    pub fn total_length(&self) -> Rational {
        let mut total = Rational::new(0, 1);
        for r in &self.ranges {
            total = total + r.length();
        }
        total
    }

    /// First time covered by any range (C++ `in()` on the first range);
    /// null rational when empty.
    pub fn first(&self) -> Rational {
        match self.ranges.first() {
            Some(r) => r.in_(),
            None => Rational::NULL,
        }
    }

    /// Frame-accurate iteration helper: snap a time to the containing
    /// frame grid of `timebase` (C++ TimeRangeListFrameIterator snap,
    /// `k_floor` rounding).
    pub fn snap(&self, time: Rational, timebase: Rational) -> Rational {
        let _ = self; // self carries no state relevant to a single snap
        rational::snap_time_to_timebase(time, timebase)
    }
}
